engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala

```diff
@@ -73,16 +77,53 @@ class CustomNodeValidator(modelData: ModelData) extends NodeDataValidator[Custom
 
   private val nodeValidator = new GenericNodeTransformationValidator(expressionCompiler, expressionEvaluator)
 
+  protected def validate(transform: SingleInputGenericNodeTransformation[_],
+                         nodeData: NodeData with WithParameters,
+                         validationContext: ValidationContext, outputVar: Option[String])(implicit metaData: MetaData): ValidationPerformed = {
+    implicit val nodeId: NodeId = NodeId(nodeData.id)
+    nodeValidator.validateNode(transform, nodeData.parameters, validationContext, outputVar) match {
+      case Valid(result) =>
+        ValidationPerformed(result.errors, Some(result.parameters))
+      case Invalid(e) => ValidationPerformed(e.toList, None)
+    }
+  }
+
+}
+
+class CustomNodeValidator(val modelData: ModelData) extends WithParametersNodeValidator[CustomNode] {
+
   override def validate(nodeData: CustomNode, validationContext: ValidationContext)(implicit metaData: MetaData): ValidationResponse = {
     val transformer = modelData.processWithObjectsDefinition.customStreamTransformers(nodeData.nodeType)._1.obj
     transformer match {
       case a:SingleInputGenericNodeTransformation[_] =>
-        implicit val nodeId: NodeId = NodeId(nodeData.id)
-        nodeValidator.validateNode(a, nodeData.parameters, validationContext, nodeData.outputVar) match {
-          case Valid(result) =>
-            ValidationPerformed(result.errors, Some(result.parameters))
-          case Invalid(e) => ValidationPerformed(e.toList, None)
-        }
+        validate(a, nodeData, validationContext, nodeData.outputVar)
+      //TODO: handle 'standard' case
+      case _ => ValidationNotPerformed
+    }
+  }
+}
+
+class SinkNodeValidator(val modelData: ModelData) extends WithParametersNodeValidator[Sink] {
```

Comment:
Hmm SinkNodeValidator and SourceNodeValidator are mostly the same.. Maybe we can do some flag to use:
```
modelData.processWithObjectsDefinition.sinkFactories(nodeData.ref.typ)._1.obj or modelData.processWithObjectsDefinition.sourceFactories(nodeData.ref.typ).obj

Reply:
Yes, but it's on purpose. The difference will emerge when we fix:
//TODO: handle 'standard' case